nexus/db-queries/src/db/datastore/region_replacement.rs

```diff
@@ -0,0 +1,886 @@
+use crate::db::error::ErrorHandler;
+use crate::db::model::Region;
+use crate::db::model::RegionReplacement;
+use crate::db::model::RegionReplacementState;
+use crate::db::model::RegionReplacementStep;
+use crate::db::model::UpstairsRepairNotification;
+use crate::db::model::UpstairsRepairNotificationType;
+use crate::db::model::VolumeRepair;
+use crate::db::update_and_check::UpdateAndCheck;
+use crate::db::update_and_check::UpdateStatus;
+use crate::db::TransactionError;
+use async_bb8_diesel::AsyncConnection;
+use async_bb8_diesel::AsyncRunQueryDsl;
+use diesel::prelude::*;
+use omicron_common::api::external::Error;
+use omicron_uuid_kinds::DownstairsRegionKind;
+use omicron_uuid_kinds::TypedUuid;
+use uuid::Uuid;
+
+impl DataStore {
+    /// Create and insert a region replacement request for a Region, returning the ID of the
+    /// request.
+    pub async fn create_region_replacement_request_for_region(
+        &self,
+        opctx: &OpContext,
+        region: &Region,
+    ) -> Result<Uuid, Error> {
+        let request = RegionReplacement::for_region(region);
+        let request_id = request.id;
+
+        self.insert_region_replacement_request(opctx, request).await?;
+
+        Ok(request_id)
+    }
+
+    /// Insert a region replacement request into the DB, also creating the
+    /// VolumeRepair record.
+    pub async fn insert_region_replacement_request(
+        &self,
+        opctx: &OpContext,
+        request: RegionReplacement,
+    ) -> Result<(), Error> {
+        self.pool_connection_authorized(opctx)
+            .await?
+            .transaction_async(|conn| async move {
+                use db::schema::region_replacement::dsl;
+                use db::schema::volume_repair::dsl as volume_repair_dsl;
+
+                diesel::insert_into(volume_repair_dsl::volume_repair)
+                    .values(VolumeRepair {
+                        volume_id: request.volume_id,
+                        repair_id: request.id,
+                    })
+                    .execute_async(&conn)
+                    .await?;
+
+                diesel::insert_into(dsl::region_replacement)
+                    .values(request)
+                    .execute_async(&conn)
+                    .await?;
+
+                Ok(())
+            })
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    pub async fn get_region_replacement_request(
+        &self,
+        opctx: &OpContext,
+        id: Uuid,
+    ) -> Result<RegionReplacement, Error> {
+        use db::schema::region_replacement::dsl;
+
+        dsl::region_replacement
+            .filter(dsl::id.eq(id))
+            .get_result_async::<RegionReplacement>(
+                &*self.pool_connection_authorized(opctx).await?,
+            )
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    pub async fn get_requested_region_replacements(
```

Comment:
how about `get_region_replacement_request_by_id`?

Reply:
So, `get_region_replacement_request` -> `get_region_replacement_request_by_id`,
Yeah, that would help.

Looking down past this, the method names below follow the same format of
`get_<STATE>_region_replacements`, so that is good.

Reply:
cool, changed in 08aacebfd